JVM IR: Bridge Lowering Fixes

sfs opened the pull request:
This PR contains large changes to `BridgeLowering`, to bring the behavior of the JVM IR backend in line with the JVM backend. There is two parts to this, which don't make much sense on their own.

The first is a change in `MethodSignatureMapper`, to compute owners using the dispatch receiver's type whenever possible. Currently we rely on fake overrides in each class and use them during codegen to pick the correct owner. This is a bit unsatisfying, since fake overrides are not really needed in the vast majority of cases and should be completely unecessary after bridge lowering. Without this change, `BrideLowering` would potentially have to insert additional fake overrides in any class containing special bridges. This is also laying the ground work for getting rid of the hack in `MethodSignatureMapper` which replaces all calls to special bridge methods except in the body of special bridge implementation methods. This could now be done as part of `BridgeLowering`, but I'll leave this change for a separate PR, since this one was getting large enough as is.

This change necessitates some fixes in other lowerings, which produced incorrect IR and that makes up for most of the commits here.

The second change is in `BridgeLowering` itself, to make it as similar as possible to the JVM backend.

Finally, I've also added some additional tests for corner cases around bridge generation. One of them is currently broken on the JS_IR backend (`compiler/testData/codegen/box/bridges/fakeOverrideThroughGenericSuperclass.kt`).

sfs commented:
Rebased on master and added code to box parameters of special super call bridges.

gogabr commented:
Overall, thanks a lot for the PR. The code of `BridgeLowering` is much clearer now than the previous version. Comments help a lot as well.

sfs commented on compiler/testData/codegen/box/bridges/addedInterfaceBridge.kt:
```diff
@@ -0,0 +1,16 @@
+
+open class C : B(), A<String>
+
+// class D should not have an additional bridge
```
That's a good point, I've replaced the test with a bytecode text test.

sfs commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt:
```diff
@@ -1,287 +1,382 @@
+            //
+            // More subtly, this check is also necessary. There is a corner-case in which we have to produce super calls
+            // for special bridge methods. Two of these special bridge methods (MutableMap.remove and getOrDefault)
+            // are @PlatformDependent and we cannot create super calls to interface methods on targets before 1.8.
```
This was unclear. I've changed the comment. :)

There are `@PlatformDependent` methods which don't exist in older JDKs and we could end up generating illegal bytecode if we tried to generate bridges for them. The point is that this optimization just so happens to subsume the check for `@PlatformDependent` methods and I wanted to document this somewhere.

It's not really that important *why* this breaks, which is what I was talking about in the comment.

sfs commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt:
```diff
@@ -1,287 +1,382 @@
         }
-
-        signaturesToSkip.add(signature)
+    private val specialBridgeMethods = SpecialBridgeMethods(context)
```
In this case it's probably fine as is, but could do with a bit of refactoring.

The `SpecialBridgeMethods` don't keep any state (just a reference to the backend context) and should be refactored into a collection of top-level functions.

The `specialBridgeCache` is just there to avoid quadratic behavior in `overriddenFinalSpecialBridges`. It would probably be cleaner to just refactor this, too.

sfs commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt:
```diff
@@ -1,287 +1,382 @@
+        }.apply {
+            copyParametersWithErasure(this@addBridge, bridge.overridden)
+            body = context.createIrBuilder(symbol).run { irExprBody(delegatingCall(this@apply, target)) }
+            overriddenSymbols += bridge.overridden.symbol
```
I think this should be fine, since the overridden symbol is for a method with the same signature as the bridge we are creating. We should get the correct return type boxing this way, and afaik this is the only reason to keep `overriddenSymbols` after bridge lowering (default arguments are already resolved).

I see two alternatives. We could add all the overridden symbols to every bridge, which might lead to too much return type boxing and could result in missing bridges. Or we could add all the overridden symbols with the same signature to the bridge as overridden symbols (and then filter out redundant overrides). The latter would be cleaner, but I'm not sure whether it is actually needed.

gogabr replied:
It does not influence anything right now, but I still prefer the second option. The reason is, IR is a language, with its own validity conditions (even if they are not written out anywhere). Every time we violate those conditions, it might bite us at some point in the future, so the less we do it, the better.

sfs replied:
True enough, that has certainly happened more than once. I've fixed the code to keep all relevant overridden symbols.

sfs commented:
Rebased on master and addressed comments. The rebase needed a fix in `AddContinuationLowering` which is in the new commit c0ddbcb.

gogabr commented:
There's been new conflicting changes on master.
Will merge after rebasing and running the tests.
Thank you!

sfs commented:
@gogabr Thanks for pointing this out, I'll just rebase and fix the conflicts

gogabr commented:
Oh, ok, thanks! (I was planning to do the rebasing myself.)

sfs commented:
Rebased and fixed conflicts.

> (I was planning to do the rebasing myself.)

Oh, I misunderstood that one. On the other hand, I needed to look at the Persistent IR changes anyway, so no problem. :)

gogabr commented:
Checks passed, merged. Thank you a lot!

gogabr commented:
FYI: After the merge, we started to get some crashes due to fake override resolution failures. Here is an example with incomplete workaround: https://github.com/JetBrains/kotlin/commit/3fdf2e737fdac396ac6bb1aaff644309cebb89c9
We are working on a fix, this message is just to notify you in case you also encounter similar crashes.

punzki commented:
@gogabr @udalov please see https://github.com/JetBrains/kotlin/pull/3032

Should also fix the same issue

udalov commented:
@punzki Unfortunately it doesn't seem like it would solve the underlying issue, it would only mask it. For example, in the `doubleDiamond.kt` test mentioned above, if you swap the supertype order in `CImpl`, I believe fake overrides will also be reversed, and now `resolveFakeOverrideWithBugCompatibility` will return `AImpl.result`, which is incorrect. The issue that we have several non-abstract fake overrides here is a bug in the JVM IR backend which needs to be fixed.

This is actually why I'm still reluctant to accept #3032 and maybe would propose to limit it to default argument functions only (via origin), since we know that that is a frontend bug and the backend should not bother with it.

punzki commented:
@udalov (Feel free to continue this conversation here or in #3032 🙂) To me `doubleDiamond.kt` also looks like a frontend bug, i.e., `CImpl` should be forced to override `get()` (ref: https://kotlinlang.org/docs/reference/classes.html#overriding-rules, `MANY_IMPL_MEMBER_NOT_IMPLEMENTED` error). Could you clarify why this is a backend bug?

udalov commented:
@punzki Basically because the implementation in `A` is overridden by the implementation in `C`. This is how override resolution works in Kotlin: first, find all non-abstract declarations that a member overrides, and then remove all declarations from that set which are overridden by at least one other declaration from that set. See [OverrideResolver. checkInheritedAndDelegatedSignatures](https://github.com/JetBrains/kotlin/blob/dc7b1fbff946d1476cc9652710df85f65664baee/compiler/frontend/src/org/jetbrains/kotlin/resolve/OverrideResolver.kt#L580).

punzki commented:
@udalov You're right; sorry I misread the inheritance chain in the test data! Let me know if you want me to revert the change in `BridgeLowering` in #3032.

sfs commented:
Isn't the `doubleDiamond` test an issue with `InheritedDefaultMethodsOnClassesLowering`? We don't generate interface implementation method bridges in `BridgeLowering` and after the interface phases we should not be seeing double multiple implementations for a concrete fake-override.

gogabr commented:
That's what I am looking at right now; `InheritedDefaultMethodsOnClassesLowering` does not generate a bridge in `CImpl` as it should.
Still, it comes down to improper implementation of `resolveFakeOverride` as it's called from that lowering as well.
